### DISCUSSION TITLE
Enhancement: Markdown support for user messages

### DISCUSSION THREAD

**Discussion by AsafMah** [danny-avila/LibreChat#1400] [Feature Requests & Suggestions]
### What happened?

When the AI answers, markdown is rendered properly.
But when the user input is shown, the markdown isn't rendered at all.
It's specifically noticeable when passing code to the ai.		

### Steps to Reproduce

1. Write in markdown
2. Wait for a response

### Screenshots

<img width="640" alt="image" src="https://github.com/danny-avila/LibreChat/assets/6424271/fa9b6a14-ff3d-4b64-97a7-9dcb7aa54348">

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by berry-13** [comment]
This is not a bug; this also exists in ChatGPT
![proof](https://github.com/danny-avila/LibreChat/assets/81851188/441e9c32-09b8-4de0-8e67-1dd6fe49a034)

**Comment by AsafMah** [comment]
Ah interesting .. before this I was using https://github.com/ztjhz/BetterChatGPT which does do it, so that explains it.

**Comment by danny-avila** [comment]
I’m converting this to a discussion because even though the implementation would be simple, this could significantly increase rendering time and slowdowns, especially on limited resource devices (mobile)

**Reply by WitnessMee** [reply]
I am not sure that I understand why this would significantly increase rendering time. The AI answers already have markdown, and they generally have a larger amount of text than the questions. (at least in my experience)

I am running the Librechat docker container of a forked repo with the changes. As you said, the implementation is indeed simple.
I updated the `isCreatedByUser ` check in _client/src/components/Chat/Messages/Content/MessageContent.tsx_ and _client/src/components/Chat/Messages/Content/Part.tsx_ and used the `Markdown` react component to parse user text.

It looks really nice and also updated it in all my previous chat messages.
In any case, keep up the amazing work on LibreChat! 🤖

**Reply by danny-avila** [reply]
I've made significant optimizations since I first commented that (Dec 20, 2023) where the effect on performance should be trivial (not null).

Feel free to make a PR!

**Reply by danny-avila** [reply]
And if not, I will get to this, this week.

**Reply by WitnessMee** [reply]
I have added a PR: https://github.com/danny-avila/LibreChat/pull/3907

**Comment by AsafMah** [comment]
Maybe it should be a setting?

**Comment by fuegovic** [comment]
I must say I'd also love to have the option to do it 
I keep using triple backticks to send codeblocks and it would look a lot nicer (and take less space visually) if the markdown was rendered on both sides

**Comment by EinMaulwurf** [comment]
I want to reignite the discussion here and provide a suggestion: How about two toggles in the settings, once for markdown rendering of user input, one for markdown rendering of the assistant's answer. That way, users who experience slowdowns can either turn it off partially or completely. 

I'm currently using a different OpenAI API interface but would like to switch to LibreChat. However, this feature (or lack there of) is holding me back, as I frequently ask ChatGPT about equations or code.

**Reply by danny-avila** [reply]
> How about two toggles in the settings, once for markdown rendering of user input, one for markdown rendering of the assistant's answer.

Great idea, I'll heavily consider it.

> I frequently ask ChatGPT about equations or code.

But even ChatGPT doesn't render markdown for user messages, does the other interface?

**Reply by EinMaulwurf** [reply]
Thanks for the reply. Yes, the interface I'm currently using is [Chatbox](https://github.com/Bin-Huang/chatbox). There, user input is rendered once the message is sent.

**Reply by Nindaleth** [reply]
[Chatbot Arena](https://chat.lmsys.org/) also renders Markdown in user input, FWIW.

**Reply by ReEnMikki** [reply]
It's not absolutely necessary, but there's also no reason to not have an option to switch it on/ off either.

**Comment by secondtruth** [comment]
This feature would be very useful for me. I'm relatively often using Markdown in prompts and rendering would therefore be optimal, especially for code blocks. The solution with two toggles are also a great idea.

**Comment by infused-kim** [comment]
I would like to add my vote for this too. I frequently send code and it would make it much easier when reading previous messages.

**Comment by insilications** [comment]
This would be extremely useful, along with with tools to format the input (e.g. add codeblocks around the selected text).

**Comment by danny-avila** [comment]
Closed by #4170
